### ISSUE TITLE
Remove all cyclic dependencies

### ISSUE THREAD

**Issue by tbosch** [angular/angular#683]
Cyclic dependencies work in some modules loaders (e.g. in SystemJS), but they don't work in CommonJS. We should get rid of them.

This issue also includes changing the build to ensure we won't get cyclic dependencies back (e.g. by walking the node.js dependencies via `module.parent` / `module.children`.

Should be done in a module by module fashion to avoid too many merge conflicts.

**Comment by pkozlowski-opensource** [comment]
Just to be precise: CommonJS does handle cyclic deps, albeit in a particular way (returning half-baked module objects): http://nodejs.org/api/modules.html#modules_cycles

In any case getting rid of cycles is a Good Thing :+1:

**Comment by rkirov** [comment]
Cycles for type annotations are very natural, for example each `View` has a `List<ViewContainer>` and each `ViewContainer` has `List<View>`. How are you planning to break that?

**Comment by mlaval** [comment]
Here is the current list of circular dependencies (generated with [madge](https://github.com/pahen/madge), which can also be used to do the validation at build time):
```
change_detection/change_detection_jit_generator -> change_detection/change_detection_util -> change_detection/exceptions -> change_detection/proto_change_detector
change_detection/change_detection_util -> change_detection/exceptions -> change_detection/proto_change_detector
change_detection/proto_change_detector -> change_detection/coalesce
change_detection/change_detection_util -> change_detection/exceptions -> change_detection/proto_change_detector -> change_detection/dynamic_change_detector
change_detection/exceptions -> change_detection/proto_change_detector -> change_detection/dynamic_change_detector
change_detection/proto_change_detector -> change_detection/dynamic_change_detector
core/compiler/element_binder -> core/compiler/view
core/compiler/shadow_dom_emulation/light_dom -> core/compiler/shadow_dom_emulation/content_tag
core/compiler/view -> core/compiler/shadow_dom_emulation/light_dom
core/compiler/view -> core/compiler/shadow_dom_emulation/light_dom -> core/compiler/view_container
core/compiler/view -> core/compiler/shadow_dom_strategy
core/compiler/view -> core/compiler/view_pool
core/compiler/pipeline/compile_control -> core/compiler/pipeline/compile_step
```

The change_detection part is quite easy to solve: extract `ProtoRecord`and the constants to a separate file.
The compiler part around the `View` will be much more challenging!

**Comment by tbosch** [comment]
So, there is actually a workaround we could use for the Types:
In `View` we don't import `ViewContainer` but include the whole module that includes `ViewContainer`. I have not tested in on NodeJS yet, but it should work in theory:
 
E.g.
```
import * as vc from './view_container';

class View {
  ...
  viewContainers: List<vc.ViewContainer>;
  ...
}
```

**Comment by mlaval** [comment]
Solved by #980
